README.md
![DeracioTSS](/misc/deractss_logo.png)

Latest Version: 1.1.1 (September 5, 2023)

![GitHub](https://img.shields.io/github/license/DERaC-IO/DeracioTSS)
![GitHub top language](https://img.shields.io/github/languages/top/DERaC-IO/DeracioTSS)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/DERaC-IO/DeracioTSS)

# Instructions
DeracioTSS is a Tornado/Python startup script. This script is primarily written for developers and affiliates of DERaC, but it is also open to the public under the [MIT License](/LICENSE). The change log is available at [CHANGELOG](/CHANGELOG.md) with some detailed discussions in [Issues](https://github.com/DERaC-IO/DeracioTSS/issues).

### Functions

- Quick Startup/Daemonization
- Routing/Handler Management (1.1.0+)
- Settings Configurations (1.1.0+)
- Server Watcher (1.1.0+)

###### Main Menu of the CLI App

```
   --- Project ---
 » 1) Health Check
   2) New Project
   --- Handler ---
   3) List Handlers
   4) Rouitng
   5) New Handler
   6) Delete Handler
   --- Setting ---
   7) Show Settings
   8) Configure
   ---------------
   9) Stop/Start Server
   0) Quit
```

##### Caveat Emptor

Scripts are coded with intensive care, but please notice issues that bugs and security holes may be found in the future. We are not responsible for any troubles caused by the usage of DeracioTSS. If you found any issues that may affect t the community, please let us know.

## Getting Started

### Startup Script

##### Installation

Download or copy'n paste `tss.py` to your server directory that the Tornado project is to be placed. For example, `wget` will work like this.

```
wget https://raw.githubusercontent.com/DERaC-IO/DeracioTSS/f6691d177de800434593cb336f102c1181c73439/tss.py
```

If you love `cURL` or `requests`, it's OK. The script requires the `questionary` package, so install it with `pip` in your virtual environment.

```
pip install questionary
```

Now you are ready to start the script with `python` or `python3`.

```
python tss.py
```

```_____________________________________________________________________
    _____                                    ______     __       __  
    /    )                           ,         /      /    )   /    )
---/----/----__---)__----__----__--------__---/-------\--------\-----
  /    /   /___) /   ) /   ) /   ' /   /   ) /         \        \    
_/____/___(___ _/_____(___(_(___ _/___(___/_/______(____/___(____/___                                                                  

Tornado Startup Script v1.1.0
Copyright (c) 2023 DERaC, LLC.
MIT License

Server is active at http://localhost:8000
```

You will see the startup logo of the program if successfully installed. Please answer questions to generate the following files. By default, folders and files are generated as shown below.

- python app file (default = `app.py`)
- template index (`index.html`)
- server startup script (default = `server.sh`)
- system daemon script (default = `tornadod.service`)

###### Directory Structure
```
<PROJECT_ROOT>
    │
    ├─ app.py
    │
    ├─ server.sh
    │
    ├─ tornadod.service
    │
    ├─ settings.py (added in 1.0.1+)
    │
    ├─ settings.d  (added in 1.0.1+)
    │    │
    │    ├─ config.json
    │    │
    │    └─ handler.json  (added in 1.0.2+)
    │
    ├─ templates
    │    │
    │    └─ index.html
    │ 
    └─ static
```

#### CLI Options (v1.0.2+)

CLI command has options whose descriptions are available with `-h` or (`--help`) option:

```
usage: tss.py [-h] [-q] [-o] [-a] [-n NAME] [-d DOMAIN] [-p PORT]

options:
  -h, --help            show this help message and exit
  -q, --quickstart      quick start with default params
  -o, --overwrite       overwrite all directories/folders
  -a, --all             generate all files
  -n NAME, --name NAME  application name (default = app)
  -d DOMAIN, --domain DOMAIN
                        port number (default = localhost)
  -p PORT, --port PORT  port number (default = 8000)
```

##### Quick-Start

The fastest option is the following that enables you to launch the Tornado server in seconds!

```
python tss.py -q -o -a
```

After installation, do `sh server.sh` on the command line.

### GitHub Cloning

```
git clone https://github.com/DERaC-IO/DeracioTSS.git
```

`git clone` will give you the copy of the entire directory of the package. If you do git-clone, paths are not detected and some other parameters are also not configured automatically, so you are requested to edit files and scripts manually that include these parameters. List of parameters are listed as follows.

###### app.py (v1.0.0)

- path to `python`
- template folder name (`__template__`)
- static file folder name (`__static__`)
- port (`__port__`)

###### app.py (v1.0.1+)

- path to `python`

###### settings.py and settings.d/config.json (v1.0.1+)

- path to `templates`
- path to `static`
- port number

###### server.sh

- `APP_NAME` (descritive app name)
- `SERVER` (path to `python`)
- `APP` (python app file)

###### tornadod.service

- `Description` (clarify service)
- `WorkingDirectory` (path to server script)
- `ExecStart` (command starting server script)
- `User` (launching user name)
- `Group` (launching group name)

### Daemonization

If installation is successful, you are already set to go. Move the service file (`tornadod.service` is default) to the system daemon directory (i.e., `/etc/systemd/system/`) in order for `systemctl` command can recognize the script. Before starting the daemon, don't forget reloading registered systems by `systemctl daemon-reload`. After that you can eventually start the daemonized Tornado application.

```
sudo mv tornadod.service /etc/systemd/system/
sudo systemctl daemon-reload
sudo systemctl start tornadod
```
If you want the Tornado application to start automatically, enable the daemon as follows (`--now` option simultaneously start the daemon after enabling it).

```
sudo systemctl enable tornadod --now
```

Now you are able to access the served website via the port you have designated, congrats!

![DeracioTSS](/misc/snapshot.png)

##### Reverse Proxy

You may want to open your Tornado web server directly to the public. In that case, set you port 80/443 (or the likes depending on your infra settings) to run the server. However, someone may feel not good to do so like me as WAF may not be applied easily or etc. In that case, you may want to use reverse proxy or CGI/FCGI techniques. Usually DERaC considers performing speed and uses reverse proxy.

## Future Roadmap

DeracioTSS tries to enhance development experiences with Tornado. The eventual goal is to build a full-stack package manager for Tornado application that is friendly even for newbie developers. The package is primarily designed for our affiliates and staffs while it is open to our communities. Following functions are going to be implemented to realize the ultimate goal (the latest version is `1.1`). Please watch [CHANGELOG](/CHANGELOG.md) for updates.

| Function | Style | Version | Status |
|:---|:---:|:---:|:---:|
|project setup script|CLI|`1.0`|Ready|
|handler (app) setup script|CLI|`1.1`|Ready|
|settings managing script|CLI|`1.2` &rarr; `1.1`|Ready|
|integrations with alembic/sqlalchemy|CLI|`1.3` &rarr; `1.2`|Working|
|routing manager|GUI/CLI|`2.0`|Working|
|handler class manager|GUI/CLI|`2.1`|Working|
|system monitor|GUI|`3.0`|-|
|wrapping web app|GUI|`3.1`|-|

##### Contributions

As an open source package, any kinds of contributions are open and welcome!

##### Donations

This package is open source, but we appreciate donations if you have felt any utility from our code and/or future development plan. Your donations via [PayPal](https://paypal.me/deracjp) or cryptos help further enhancements/developments, thanks!

| Crypto | Address |
|:---:|:---:|
| BTC | ![BTC](/misc/btc_bitflyer_qr.png)<br>3GSGPvNeSv1j9LobjzZWiCL2Vd29rpmyXE |
| ETH | ![ETH](/misc/eth_bitflyer_qr.png)<br>0x8928534c8beca7875059edce7afae202836a0d4c |

## About DERaC

Decentralized Economy Research & Consulting (DERaC) was founded by Tetsuya (Tedd) Saito, Ph.D. in December 2022 based on his specilties -- network security, mathematical/data science, economics, and 25+ years of experience in web development. DERaC provides services on network security, web development, statistical analysis, mathematical modeling, economic consulting, and web development with several affiliated specialists of each field.

Occasional job opportunities are posted on our website at https://derac.io (en) or https://derac.jp (ja), but applications from talented specialists/developers that are capable of Python, R, statistics, economics/econometrics, applied mathematics, machine learning, and ethical hackers are always welcome.


CHANGELOG.md
![DeracioTSS](/misc/deractss_logo.png)

# Change Log

## Version 1.1.x

### 1.1.1

#### Implemented Enhancement(s)

- None

#### Fixued Bug(s)

- Error is fixed when no handler is registered in `DeracioTSS.health()` ([Issue #6](https://github.com/DERaC-IO/DeracioTSS/issues/6))

#### Housekeeping Change(s)

- None

### 1.1.0

#### Implemented Enhancement(s)

1. Routing/handler CLI manager
2. Settings CLI manager
3. System health checker (primitive)
4. Server stop/start switch

#### Fixued Bug(s)

- None

#### Housekeeping Change(s)

1. Opening menu
2. Much faster quick-starter options
3. Rationalizing questionaries

## Version 1.0.x

### 1.0.3

#### Implemented Enhancement(s)

- None

#### Fixued Bug(s)

1. An error to generate the shell script path while quick-starting has been fixed

#### Housekeeping Change(s)

- None

### 1.0.2

#### Implemented Enhancement(s)

1. Modified to manage route/handler in JSON file (`handler.json`) in `settings.d`
2. CLI options are added for quick-start

#### Fixued Bug(s)

- None

#### Housekeeping Change(s)

1. Revisions of asking questions including the one mentioned at [Issue #4](https://github.com/DERaC-IO/DeracioTSS/issues/4)
2. Re-coded as class instead of single function
3. Text decorations for CLI outputs

### 1.0.1

#### Implemented Enhancement(s)

1. Generating settings.py from JSON files in `settings.d` ([Issue #3](https://github.com/DERaC-IO/DeracioTSS/issues/3))

#### Fixued Bug(s)

- None

#### Housekeeping Change(s)

1. `version` parameter is added in `main` to display its version


tornado_project/app.py
#!/home/ecs-user/japorn/.pypy/bin/python
# -*- coding: utf-8 -*-
import tornado.ioloop, tornado.web
import os, re, json, threading

# Define fundamental parameters for the service
# These params can be alternatively in, i.e., settings.py
__port__ = 8000
__templates__ = "templates"
__static__ = "static"

# Classes are separately stored in external files
# Follow the structure rule of each project
class MainHandler(tornado.web.RequestHandler):
    def get(self):
        params = {
            'title': 'Deracio Tornado/Python Setupper'
        }
        self.render("index.html", params = params)

# Definition for the web service
# Routing, debug (autoload) setting, template/static folders, etc...
application = tornado.web.Application(
    [
        ("/", MainHandler),
    ],
    debug=True,
    template_path=os.path.join(os.getcwd(), __templates__),
    static_path=os.path.join(os.getcwd(), __static__),
)

# Startup script for the server (read by server.sh)
if __name__ == "__main__":
    application.listen(__port__)
    tornado.ioloop.IOLoop.instance().start()


versions/tss-1.0.1.py
"""
DeracioTSS -- Tornado Startup Script
Copyright (c) 2023 DERaC, LLC.

BTC: 3GSGPvNeSv1j9LobjzZWiCL2Vd29rpmyXE
ETH: 0x8928534c8beca7875059edce7afae202836a0d4c
"""
# -*- coding: utf-8 -*-
import os, sys, re, json, subprocess, questionary

def main():
    version = '1.0.1'
    banner = """
_____________________________________________________________________
    _____                                    ______     __       __  
    /    )                           ,         /      /    )   /    )
---/----/----__---)__----__----__--------__---/-------\--------\-----
  /    /   /___) /   ) /   ) /   ' /   /   ) /         \        \    
_/____/___(___ _/_____(___(_(___ _/___(___/_/______(____/___(____/___                                                                  
"""
    print(banner)
    print(f'Tornado Startup Script v{version}')
    print('Copyright (c) 2023 DERaC, LLC.')
    print('MIT License\n')
    if questionary.confirm('Starting a new project?').ask():
        name = "app"
        name = questionary.text(f'Let me know the project name', default=name).ask()
        if name == '':
            print('\nProject name cannot be blank...\n')
            exit()
        name = re.sub('\s', '_', name)
        if questionary.confirm(f'Create {name}?').ask():
            template = 'templates'
            static = 'static'
            port = '8000'
            app_name = f'{name}.py'
            if not questionary.confirm(f'Use default template folder name ({template})?').ask():
                template = questionary.text(f'Template folder name').ask()
            if not questionary.confirm(f'Use default static file folder name ({static})?').ask():
                static = questionary.text(f'Static file folder name').ask()
            port = questionary.text(f'Which port will you use?', default=port).ask()
            if not questionary.confirm(f'Use default app file name ({app_name})?').ask():
                app_name = questionary.text(f'App file name').ask()
            python_path = subprocess.run("which python3", shell=True, stdout=subprocess.PIPE, encoding='utf-8')
            python_path = questionary.path(f'Path to python3', default=python_path.stdout.strip()).ask()
            
            print('\nStart creating files...')

            exist = os.path.isfile(app_name)
            if exist:
                overwrite = questionary.confirm(f'Overwrite existing {app_name}?', default=False).ask()
            if not exist or overwrite:
                script = f'#!{python_path}\n'
                script += """
# -*- coding: utf-8 -*-
import tornado.ioloop, tornado.web
import os, re, json, threading

# Define fundamental parameters for the service
# These params are in settings.py generated from files in settings.d

from settings import config

__template__ = config.get("template")
__static__ = config.get("static")
__port__ = config.get("port")

# Classes are separately stored in external files
# Follow the structure rule of each project
class MainHandler(tornado.web.RequestHandler):
    def get(self):
        params = {
            'title': 'Deracio Tornado/Python Setupper'
        }
        self.render("index.html", params = params)

# Definition for the web service
# Routing, debug (autoload) setting, template/static folders, etc...
application = tornado.web.Application(
    [
        ("/", MainHandler),
    ],
    debug=True,
    template_path=os.path.join(os.getcwd(), __template__),
    static_path=os.path.join(os.getcwd(), __static__),
)

# Startup script for the server (read by server.sh)
if __name__ == "__main__":
    application.listen(__port__)
    tornado.ioloop.IOLoop.instance().start()
"""
                try:
                    with open(app_name, 'w') as f:
                        f.write(script)
                        f.close()
                    print(f'Folder {app_name} is generated successfully!')
                except Exception as e:
                    print(e)
                    exit()
            else:
                print(f'{app_name} already exists...')
            
            print('\nCreating settings files...')
            
            if not os.path.isdir('settings.d'):
                os.mkdir('settings.d')
            exist = os.path.isfile('settings.py')
            if exist:
                overwrite = questionary.confirm(f'Overwrite existing settings.py?', default=False).ask()
            if not exist or overwrite:
                config = {
                    'template': template,
                    'static': static,
                    'port': int(port),
                }
                with open('settings.d/config.json', 'w') as f:
                    f.write(json.dumps(config))
                    f.close()
                
                files = os.listdir("settings.d")
                script = "# Settings generated from JSON files (identified by .json) in settings.d folders\n\n"
                for file in files:
                    name = file.split('.')
                    if name[1] == 'json':
                        script += name[0] + " = {\n"
                        with open(f"settings.d/{file}") as f:
                            entry = f.read()
                            if entry:
                                entry = json.loads(entry)
                                for key, value in entry.items():
                                    if type(value) is str:
                                        script += '\t"{}": "{}",\n'.format(key, value)
                                    else:
                                        script += '\t"{}": {},\n'.format(key, value)
                            f.close()
                        script += "}\n"
                    with open('settings.py', 'w') as f:
                        f.write(script)
                        f.close()

            print('\nCreating template folder...')

            if not os.path.isdir(template):
                try:
                    os.mkdir(template)
                    print(f'Folder {template} is created successfully!')
                except Exception as e:
                    print(e)
                    exit()
            else:
                print(f'Folder {template} already exists...')

            print()
            
            exist = os.path.isfile(f'{template}/index.html')
            if exist:
                overwrite = questionary.confirm(f'Overwrite existing {template}/index.html?', default=False).ask()
            if not exist or overwrite:
                script ="""
<!DOCTYPE html>
<html lang='en'>
<head>
    <meta charset="utf-8">
    <title>{{ params.get('title') }}</title>
    <!-- UIkit CSS -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/uikit@3.16.24/dist/css/uikit.min.css" />
    <!-- UIkit JS -->
    <script async src="https://cdn.jsdelivr.net/npm/uikit@3.16.24/dist/js/uikit.min.js" defer></script>
    <script async src="https://cdn.jsdelivr.net/npm/uikit@3.16.24/dist/js/uikit-icons.min.js" defer></script>
</head>
<body>
{% block body %}
<div class="uk-margin-top uk-text-center">
    <img src="https://raw.githubusercontent.com/DERaC-IO/DeracioTSS/2a43aac2a5138902f83d24b38a367b7366aa8548/misc/deractss_logo.png">
    <div class="uk-h1">Congratulations!</div>
    <div class="uk-text-success">Tornado is now working!</div>
    <div class="uk-h2">Donations</div>
    <center>
    <p class="uk-width-1-3@m">
        DeracioTSS is open source under MIT license, but we appreciate donations if you have felt any utility from our code!
    </p>
    <table class="uk-table uk-width-1-3@m uk-text-center uk-margin-top">
        <tr><td>BTC</td><td>ETH</td></tr>
        <tr>
            <td>
                <img src="https://raw.githubusercontent.com/DERaC-IO/DeracioTSS/2a43aac2a5138902f83d24b38a367b7366aa8548/misc/btc_bitflyer_qr.png">
            </td>
            <td>
                <img src="https://raw.githubusercontent.com/DERaC-IO/DeracioTSS/2a43aac2a5138902f83d24b38a367b7366aa8548/misc/eth_bitflyer_qr.png">
            </td>
        </tr>
    </table>
    </center>
    <div class="uk-text-primary">For BTC: 3GSGPvNeSv1j9LobjzZWiCL2Vd29rpmyXE</div>
    <div class="uk-text-primary">For ETH: 0x8928534c8beca7875059edce7afae202836a0d4c</div>
</div>
{% end %}
</body>
</html>
"""
                try:
                    with open(f'{template}/index.html', 'w') as f:
                        f.write(script)
                        f.close()
                    print(f'{template}/index.html is generated successfully!')
                except Exception as e:
                    print(e)
                    exit()
            else:
                print(f'{template}/index.html already exists...')

            print('\nCreating static file folder...')

            if not os.path.isdir(static):
                try:
                    os.mkdir(static)
                    print(f'Folder {static} is created successfully!')
                except Exception as e:
                    print(e)
                    exit()
            else:
                print(f'Folder {static} already exists...')

            print()

            if questionary.confirm('Do you wish to generate server startup script?').ask():
                shell = questionary.select(
                    'Which shell do you use?',
                    choices = ['bash','zsh','csh','tcsh','sh'],
                    default = 'bash'
                ).ask()
                shell_path = subprocess.run(f"which {shell}", shell=True, stdout=subprocess.PIPE, encoding='utf-8')
                shell_path = questionary.path(f'Path to {shell}', default=shell_path.stdout.strip()).ask()
                server_name = questionary.text(f'Script name', default="server.sh").ask()
                mode = questionary.text('File permission mode', default='0755').ask()
                exist = os.path.isfile(server_name)
                if exist:
                    overwrite = questionary.confirm(f'Overwrite existing {server_name}?', default=False).ask()
                if not exist or overwrite:
                    script = f'#!{shell_path}\n'
                    script += """
# Define the app name
APP_NAME="Deracio Tornado/Python"

# Full path to the app directory
HOME_DIR=`pwd`

# Full path to the server program
# Tornado launches webserver of itself => path to python
"""                 
                    script += f'SERVER="{python_path}"\n'
                    script += """
# App file name that includes 'application' to 'listen'
### For example:
### application = tornado.web.Application(...)
### application.listen(...)
"""
                    script += f'APP="{ app_name }"\n'
                    script += """
# Startup script
echo "Starting $APP_NAME"
$SERVER $HOME_DIR/$APP
"""
                    try:
                        with open(server_name, 'w') as f:
                            f.write(script)
                            f.close()
                        print(f'\n{ server_name } is generated successfully!')
                        if subprocess.call(["chmod", mode, server_name]) == 0:
                            print(f'Permission is set {mode} for { server_name }\n')
                        
                    except Exception as e:
                        print(e)
                        exit()
                else:
                    print(f'{ server_name } already exists...\n')

                if questionary.confirm('Do you wish to generate system daemon script?').ask():
                    daemon_name = questionary.text('System daemon file name', default='tornadod.service').ask()
                    description = questionary.text('Brief system description', default='Tornado Web Service').ask()
                    cwd = questionary.path('Working directory', default=os.getcwd()).ask()
                    user = questionary.text('User name to execute', default='nobody').ask()
                    group = questionary.text('Group name to execute', default='nobody').ask()
                    mode = questionary.text('File permission mode', default='0755').ask()

                    exist = os.path.isfile(daemon_name)
                    if exist:
                        overwrite = questionary.confirm(f'Overwrite existing {daemon_name}?', default=False).ask()
                    if not exist or overwrite:
                        script = """
[Unit]
# Clarify to descrive this service
"""
                        script += f'Description={description}\n'
                        script += """
[Service]
# Full path to your project
"""
                        script += f'WorkingDirectory={cwd}\n'
                        script += '# Full path to your server startup script\n'
                        script += f'ExecStart="{cwd}/{server_name}"'
                        script += """
Restart=always
RestartSec=3s
Type=simple

# User and group names to launch the service
# Avoid root or equivalent user/group
"""
                        script += f'User={user}\n'
                        script += f'Group={group}\n'
                        script += """
[Install]
WantedBy=multi-user.target
"""
                        try:
                            with open(daemon_name, 'w') as f:
                                f.write(script)
                                f.close()
                            print(f'{ daemon_name } is generated successfully!')
                            mode = '0755'
                            if subprocess.call(["chmod", mode, daemon_name]) == 0:
                                print(f'Permission is set {mode} for { daemon_name }')
                        except Exception as e:
                            print(e)
                            exit()
                        finally:
                            print(f'\nFollow the procedure below to daemonize { daemon_name }')
                            print(f'1. sudo mv { daemon_name } /etc/systemd/system/')
                            print(f'2. sudo systemctl daemon-reload')
                            print(f'3. sudo systemctl enable { daemon_name } --now')
        print('\nCongratulations!')
        print('Tornado is ready to work with you now!')
        print('Don\'t forget to "pip install tornado"')
    print('\nDonations are welcome!\n')
    print('BTC: 3GSGPvNeSv1j9LobjzZWiCL2Vd29rpmyXE\n')
    print('ETH: 0x8928534c8beca7875059edce7afae202836a0d4c\n')
    print('bye...\n')

if __name__=='__main__':
    main()
